update: add displayName and password description，icons in setting page

SeijaHuang opened the pull request:
Added
- refactor setting pages with fully utilizing react-hook-form,
- added account/setting folders under module to modularize setting page
- added setting-form.component.tsx containing the setting page component
- added SettingFormLayout containing the form layout of this page
- added SettingFormGroup containing the main form part of this page
- added FormInput to both control password input and other input fields
- added FormIcon to control different icons respond to different fields
- updated test file in align with new code

Refacor
- refactored login form that follows a consistent naming convention
Resolve CP-83

screenshot:
<img width="1604" alt="image" src="https://github.com/user-attachments/assets/9100231e-0121-4386-b2b3-79ee2e547008">
<img width="1631" alt="image" src="https://github.com/user-attachments/assets/5859ca54-edce-4445-ba32-f6b981e22d74">


unit test result:
<img width="1016" alt="image" src="https://github.com/user-attachments/assets/ad26999e-bd02-41d3-b8b1-69f8fa09ac62">

how to test:
1. login first
2. in the dashboard page click the avatar to show the dropdown, and then click the setting to navigate to the setting page
3. follow the instructions prompt during your test

Chris-Wang commented on src/app/(protected)/account/settings/page.tsx:
```diff
@@ -52,8 +85,15 @@ const Setting = () => {
                 <FormGroup>
+                  <Description>
+                    {isFocused.displayName &&
+                      'Display name should between 4 to 15 characters, only letters, numbers, dashes and underscore are allowed'}
```
we probably think of abstracting this magic string, and make it shareable between different pages

SeijaHuang replied:
add share styled component
<img width="871" alt="image" src="https://github.com/user-attachments/assets/3c3690fe-3f8a-4444-a390-381eba5ac64b">

Chris-Wang commented on src/app/(protected)/account/settings/page.tsx:
```diff
@@ -85,21 +130,26 @@ const Setting = () => {
                 <FormGroup>
+                  <Description>
+                    {isFocused.password &&
+                      'Password should be 8 to 32 characters, including letter, number and special character'}
```
we probably think of abstract this magic string, and make it shareable between different pages

SeijaHuang replied:
add share styled component

NorrisWu0 commented on src/app/(protected)/account/settings/page.tsx:
```diff
@@ -1,160 +1,5 @@
-    </Container>
-  );
-};
+const Setting = () => <SettingForm />;
```
Can `export default SettingForm` work?

SeijaHuang replied:
Yes! Reafactored!

SeijaHuang replied:
refactor as export default function SettingPage(){}

NorrisWu0 replied:
not exactly what I meant, but I'm okay (happier even) with this 👌

NorrisWu0 commented on src/module/account/settings/SettingFormGroup.tsx:
```diff
@@ -0,0 +1,55 @@
+import { Controller, useFormContext } from 'react-hook-form';
+import FormInput from './_component/FormInput/FormInput';
+
+export const SettingFormGroup = () => {
```
Why break this out?

SeijaHuang replied:
nothing special

NorrisWu0 replied:
what do you mean nothing special? hahah.

If we not reusing this then maybe it can be part of the actual setting form? or if it's for organisation purpose, I reckon we can have it colocate with the actual form?

SeijaHuang replied:
Sorry. Misunderstand break this out. To make SettingFormGroup a component is in alignment with Damon's example of the login page in which all the Controllers(form group in the original file) are extracted to become a component called FormGroup.

99032000 replied:
I extracted a component for the form group because there were so many wrappers on the controller, which made the code look messy.  You built these reusable components to make it much better,  I think you just put these codes back on the same page in a function should be ok.  

what do you think @SeijaHuang  @NorrisWu0 ?

NorrisWu0 replied:
> put these codes back on the same page in a function

wdym by this?

I guess my idea is 

```tsx

export function SettingForm() {
  const { ..... , control} = useForm(...)
  return (
    <form>
      ...
      <SettingFormGroup control={control}/>
      ...
    </for>
}

function SettingFormGroup({ control }) {
  return (
    <inputs....../>
  )
}
```

oh and, why do we use form context here, could we just pass the `control` as a props? @SeijaHuang

SeijaHuang replied:
sure, I will refactor here, merging the form group component into the layout. I think that will be more straightforward and will not use form context anymore.

NorrisWu0 commented on src/module/account/settings/_component/FormInput/FormInput.tsx:
```diff
@@ -0,0 +1,72 @@
+      <FormGroupLabel>{placeholder}</FormGroupLabel>
+      <FormGroupField>
+        <FormIcon name={name} />
+        {name === 'password' && (
```
Best to put them as separate components instead of switching them based on `name`

SeijaHuang replied:
Okkkkk

NorrisWu0 commented on src/module/account/settings/_component/FormInput/FormInput.tsx:
```diff
@@ -0,0 +1,72 @@
+import styled from 'styled-components';
+import { FieldErrors } from 'react-hook-form';
+import EyeIcon from 'mdi-react/EyeIcon';
+import FormIcon from './_component/FormIcon';
```
Whats FormIcom for? Maybe Eyey icon can be part of it?

SeijaHuang replied:
Okkkkk, add Eyey icon into FormIcon Component

SeijaHuang replied:
actually FormIcon has icon wrapper, EyeyIcon has no wrapper

NorrisWu0 replied:
okay... sounds complicated... not gonna dig into this

NorrisWu0 commented on src/module/account/settings/_component/FormInput/_component/FormIcon.tsx:
```diff
@@ -0,0 +1,24 @@
+
+function FormIcon({ name }: { name: keyof typeof iconForField }): React.JSX.Element {
+  return (
+    <>
```
Don't need react fragment here

SeijaHuang replied:
deleted

NorrisWu0 commented on src/module/account/settings/setting-form.component.test.tsx:
```diff
@@ -0,0 +1,57 @@
+import { render, screen } from '@testing-library/react';
```
不看

NorrisWu0 commented on src/module/account/settings/setting-form.component.tsx:
```diff
@@ -0,0 +1,26 @@
+import SettingFormLayout from './SettingFormLayout';
+
+const SettingForm = () => (
+  <Container>
```
不看！！

NorrisWu0 commented on src/module/auth/login-form/FormLayout.tsx:
```diff
@@ -0,0 +1,88 @@
+import { USER_LOGIN } from '@/graphql/auth';
+import LoginFormGroup from './LoginFormGroup';
+
+type loginData = { email: string; password: string; remember_me: boolean };
```
type always Pascalcase

SeijaHuang replied:
refactored

NorrisWu0 commented on src/module/auth/login-form/FormLayout.tsx:
```diff
@@ -0,0 +1,88 @@
+
+  useEffect(() => {
+    if (rememberMe !== undefined && typeof window !== 'undefined') {
+      localStorage.setItem(REMEMBER_ME, rememberMe.toString());
```
Why not do this when form submit? Seems a little bit off to have this as an effort

SeijaHuang replied:
This is merged from dev= = really don't konw

NorrisWu0 replied:
okie, let's keep it here for now

NorrisWu0 commented on src/module/auth/login-form/FormLayout.tsx:
```diff
@@ -0,0 +1,88 @@
+      }
+      if (originUrl && originUrl !== '/') {
+        // history.push(originUrl);
+        router.back();
```
What's this for again?

SeijaHuang replied:
This is merged from dev= = really don't konw

NorrisWu0 replied:
koool, keep it here for now as well

NorrisWu0 commented on src/module/account/settings/SettingFormLayout.tsx:
```diff
@@ -0,0 +1,53 @@
+  return (
+    <FormProvider {...methods}>
+      <FormContainer $horizontal onSubmit={handleSubmit(onSubmit)}>
+        <SettingFormGroup />
```
so by colocate => I mean to having `<SettingFormGroup/>` in this file

SeijaHuang replied:
Sorry. Misunderstand break this out. To make SettingFormGroup a component is in alignment with Damon's example of the login page in which all the Controllers(form group in the original file) are extracted to become a component called FormGroup.

NorrisWu0 commented on src/module/account/settings/_component/FormInput/FormPasswordInput.tsx:
```diff
@@ -0,0 +1,56 @@
+          type="button"
+          onClick={() => setPasswordShown((shown) => !shown)}
+        >
+          <EyeIcon />
```
btw, a better ux would be to show a open eye icon 🙉 when showing password and a close eye icon 🙈 when not showing password

SeijaHuang replied:
Yes, I totally agree with you! And this improvement is refining in the CP-84
<img width="289" alt="image" src="https://github.com/user-attachments/assets/f065b00c-7545-4354-b0e4-da136692f475">

NorrisWu0 replied:
ahhh okay,

NorrisWu0 commented on src/module/auth/login-form/login-form.component.tsx:
```diff
@@ -0,0 +1,54 @@
+import FormLayout from './FormLayout';
+export default function LoginForm() {
+  return (
+    <AccountWrap>
```
不看

NorrisWu0 commented on src/shared/utils/helpers.ts:
```diff
@@ -1,3 +1,30 @@
+import * as z from 'zod';
+
+export const userSchema = z.object({
```
okay, so looking at the object, it looks like it's validating a form? but variable name `userSchema` sounds like it's validating a user object?

also, is this validating one specific form or multiple? I would co-locate this with the form if it's just for one form.

NorrisWu0 commented:
@SeijaHuang congrats! you got conflicts to resolve! I merge `dev` first then make the changes.

NorrisWu0 commented on src/module/account/settings/SettingFormLayout.tsx:
```diff
@@ -0,0 +1,78 @@
+import * as z from 'zod';
+import { SettingFormGroup } from './SettingFormGroup';
+
+const userSchema = z.object({
```
being extremely picky here, maybe just call this `schema` or `formSchema`, or even `registerFormSchema` if you want to be explicit.

SeijaHuang replied:
okkk

SeijaHuang committed:
update: add displayName and password description and icons in setting page in setting page,

Added
- refactor setting pages with fully utilizing react-hook-form,
- added account/setting folders under module to modularize setting page
- added setting-form.component.tsx containing the setting page component
- added SettingFormLayout containing the form layout of this page
- added SettingFormGroup containing the main form part of this page
- added FormInput to both control password input and other input fields
- added FormIcon to control different icons respond to different fields
- updated test file in align with new code

Refacor
- refactored login form that follows a consistent naming convention
Resolve CP-83

Chris-Wang commented on src/module/account/settings/SettingFormLayout.tsx:
I just feel a bit awkward  here, why the file name staring with capital letter?

SeijaHuang replied:
The file name will follow the rule that it should start with capital letter if it is a tsx file.

Chris-Wang commented on src/module/auth/login-form/FormLayout.tsx:
same here, why the file name staring with capital letter?

SeijaHuang replied:
The file name will follow the rule that it should start with capital letter if it is a tsx file.

Chris-Wang commented on src/shared/components/form/Error.tsx:
same here, why the file name staring with capital letter?

SeijaHuang replied:
The file name will follow the rule that it should start with capital letter if it is a tsx file.

Chris-Wang commented on src/shared/components/form/FormElements.tsx:
same here, why the file name staring with capital letter?

SeijaHuang replied:
The file name will follow the rule that it should start with capital letter if it is a tsx file.

Chris-Wang commented on src/module/account/settings/SettingFormLayout.tsx:
```diff
@@ -0,0 +1,117 @@
+const formSchema = z.object({
+  displayName: z
+    .string()
+    .min(1, { message: 'Display name is required' })
```
be careful the magic string here, since you're gonna take the translation ticket, I won't ask for a change here, these string will be further extracted to somewhere else in that task

SeijaHuang replied:
added error messages enum in the shared/utils/helpers

Chris-Wang commented on src/module/account/settings/setting-form.component.test.tsx:
```diff
@@ -0,0 +1,57 @@
+      )
+    ).toBeVisible();
+
+    await userEvent.type(displayNameInput, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
```
we tried to mock the input data as close to the as real world case as possible, please try to use something looks better

SeijaHuang replied:
changed the mocked password closer to real word case as 'englishisnotdifficulttolearn1234567890!@#butchineseisveryhard'

Chris-Wang commented on src/module/account/settings/setting-form.component.test.tsx:
```diff
@@ -0,0 +1,57 @@
+    ).toBeVisible();
+    expect(
+      await screen.findByText(
+        /Password must contain at least one letter, one number and one special character/i
```
if the magic string was extracted somewhere else, you don't need to copy and paste the whole string here

SeijaHuang replied:
added error messages enum in the shared/utils/helpers

Chris-Wang commented on src/module/account/settings/setting-form.component.test.tsx:
```diff
@@ -0,0 +1,57 @@
+    expect(await screen.findByText(/password is required/i)).toBeVisible();
+  });
+
+  it('should show invalid field errors for each invalid input field', async () => {
```
please, break this test down into smaller pieces, so far I can see, it tested for different inputs and covered:
- input length
- valid email
- input format

SeijaHuang replied:
Sure. Break tests into max length test, min length test, invalid test

Chris-Wang commented on src/module/account/settings/setting-form.component.test.tsx:
```diff
@@ -0,0 +1,57 @@
+  render(<SettingForm />);
+}
+
+describe('setting', () => {
```
shouldn't there be a test for happy path?

SeijaHuang replied:
Add a todo reminder here

Chris-Wang commented on src/module/auth/login-form/FormLayout.tsx:
```diff
@@ -1,18 +1,18 @@
 import LoginFormGroup from './LoginFormGroup';
 
-type loginData = { email: string; password: string; remember_me: boolean };
+type LoginData = { email: string; password: string; remember_me: boolean };
```
good catch!

Chris-Wang reviewed (approved):
LGTM
